unit tests running with windows shared dll

AndreasSchoenle opened the pull request:
Changes mainly to CMake files addressing the following issues

- generated definitions stored in the source tree is incompatible with multiple build configurations using the same source checkout. Moved to binary directory
- when using DBUG instead of DEBUG, some compiler errors needed to be fixed
- when compiling g3logger as DLL it should use the runtime DLLs. Otherwise there is problems with allocators (inside the DLL) using a different heap than the deallocators (outside the DLL, i.e. in the test programs). Changed compiler flags to /MD, /MDd to fix this.
- compiler flags where overwritten instead of just adding the /MT, /MTd 
- even with GNU binutils installed, the git version extraction did not work for me due to single quotation marks not being accepted
- needed to set the working directory for the git command as my build directory is outside the source directory.

KjellKod commented:
There's an issue raised for the DBUG usage. The summary in this pull request mentions that it is fixed but I don't see it

KjellKod commented:
Overall the pull request seems to target several or all of the recent Windows issues that were raised. 

The summary mentions them but I do not see them addressed in the pull request so I'm assuming it is still in-progress.

Please holler when it is ready for re-review

AndreasSchoenle commented:
I believe it is all there (now):
- The DBUG usage is in loglevels.cpp and test_unit/test_message.cpp and test_unit/tester_sharedlib.cpp where the DEBUG symbol is used whether or not the DBUG option is on. 
- The /MD vs /MT issue and generated definition is only in CMake
- I added a correction to the generated definitions file (appended it to HEADER_FILES) since your comment so it will be installed with the rest of the headers. Otherwise the change is merely in GenerateMacroDefinitionsFile.cmake writing it to the CMAKE_BINARY_DIR instead of the source tree.
- The git command call to test for the version and build number works for me under windows but the working-directory change for linux out-of-source builds is untested.
Please let me know if anything 
else is still missing? 

Please let me know if there is still something missing.

AndreasSchoenle commented:
One more comment: Would you mind if I changed all the message( ... ) statements that do not represent warnings in the CMake files to message( STATUS ... )? When running the CMake GUI it actually helps a lot to spot the actual errors when the CMake run fails?

Also: I have added the wchar support again in my repo. It is in the wchar-support branch but I have merged the changes in this pull request (as I could not have tested without them). If it is ok, I will re-create a pull request, when we are done with this one?
Otherwise I can try to base one on master and test without switching to DBUG and with g3logger as static library for now (otherwise all unit tests complain about the memory thing mentioned above) - but I would prefer to do this sequentially.

Also: I have successfully compiled 1.3 with gcc 4.7 on our embedded platform. Only needed to change -std=c++14 back to -std=c++11 and add the other platform specific compiler flags. So if we successfully merge wchar support this time I will definitely switch to 1.3 in production.

AndreasSchoenle commented:
You mentioned bringing back the static library option. Should I add an option for this? I had removed the shared option because it seemed dead - do you want to keep this option name and bring it back to life? Also: Do I understand correctly that you would want to NOT have both libraries built but either static or DLL?
About the Runtime: IMHO it would still be preferable if the library was linking to the DLL runtime (at least as default) as it seems to be the default for almost all windows DLLs and programs. If g3logger is a static lib and links to the static Runtime it forces all modules using it to do the same to avoid aforementioned heap trouble. 

Except for the dynamic/static thing: Do you still see issues with the PR? Any thoughts on the questions from my previous comment?

AndreasSchoenle commented:
Added G3_SHARED_LIB option as suggested in #174.

AndreasSchoenle commented:
OK, seems like I got it together. Appveyor is now testing with DEBUG => DBUG and dynamic logging on (hope this is ok) and runs the unit tests in windows.
It seems like the g3log-FATAL-contract.exe exits with code 3 by design, so I have changed the call such that this does not result in a failed test. 

I have merged all changes to master since then.

Please let me know if anything is missing.

AndreasSchoenle commented:
I beleive the segfault is not related to the pull request. There seems to be a race condition in pushMessageToLogger() where after the call to isLoggingInitialized() g_logger_instance is used without holding a lock to g_logging_init_mutex(). 

Not sure how to fix without impacting performance. Probably another boolean that is set when testing for initialization which defers destruction?

KjellKod commented on appveyor.yml:
```diff
@@ -28,11 +28,11 @@ before_build:
-  - cd c:\projects\g3log\build\Release
-  - g3log-FATAL-contract.exe
+  - cmd /c Release\g3log-FATAL-contract.exe || exit /B 0
+  - msbuild RUN_TESTS.vcxproj /p:Configuration=Release
```
Cool thanks.

When this is working later my hope is that appveyor will run all the unit tests. 

Thanks for the DEBUG to DBUG change too.  Great to capture this on this CI

KjellKod commented on src/loglevels.cpp:
```diff
@@ -19,7 +19,11 @@ namespace g3 {
-         {g3::kDebugValue, {DEBUG}},
+#if (defined(CHANGE_G3LOG_DEBUG_TO_DBUG))
+		  { DBUG.value, { DBUG } },
+#else
```
It's cramming too much into a single pull request which might mean it's take longer time to get it in but I truly appreciate that you took the time to address this issue

Thanks

AndreasSchoenle replied:
OK, should I factor the DEBUG vs DBUG issue out?

KjellKod commented on test_unit/test_message.cpp:
```diff
@@ -189,21 +189,26 @@ TEST(Message, localtime_formatted) {
 
+#if (defined(CHANGE_G3LOG_DEBUG_TO_DBUG))
+#define DEBUG_NAME DBUG
+#else
```
This would be better to keep in the generated options file. Currently you are repeating it in multiple places which should not be needed as long as they are including the generated header file

KjellKod commented on test_unit/tester_sharedlib.cpp:
```diff
@@ -11,6 +11,12 @@
 
+#if (defined(CHANGE_G3LOG_DEBUG_TO_DBUG))
+#define DEBUG_NAME DBUG
+#else
```
Please remove and use the generated header file instead (see comment above)

AndreasSchoenle replied:
OK, will do.

KjellKod commented:
The crash is related to this issue. https://github.com/KjellKod/g3log/issues/167

Once the top priority issues are dealt with I'm taking this one on next. Optionally you could add the great disabled test to the test to avoid it since it is already raised as an issue. https://github.com/KjellKod/g3log/issues/167

AndreasSchoenle commented on Build.cmake:
```diff
@@ -93,20 +94,32 @@ ENDIF()
    file(GLOB SRC_FILES ${LOG_SRC}/g3log/*.h ${LOG_SRC}/g3log/*.hpp ${LOG_SRC}/*.cpp ${LOG_SRC}/*.ipp)
    file(GLOB HEADER_FILES ${LOG_SRC}/g3log/*.hpp ${LOG_SRC}/*.hpp)
    
+   list( APPEND HEADER_FILES ${GENERATED_G3_DEFINITIONS} )
```
This is only needed so the file shows up in IDEs (SRC_FILES) and to make sure it is installed later where you use HEADER_FILES to install the includes.

AndreasSchoenle commented on Build.cmake:
```diff
@@ -60,8 +61,8 @@ ELSEIF(${CMAKE_CXX_COMPILER_ID} STREQUAL "GNU")
    set(PLATFORM_LINK_LIBRIES dbghelp)
-   set(CMAKE_CXX_FLAGS_RELEASE "/MT")
-   set(CMAKE_CXX_FLAGS_DEBUG "/MTd")
+   # set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} /MD")
```
Under windows there is no scenario I see where it would be desirable to link against the static Runtime. The DLL runtime is the default, so I removed these flags. As mentioned above using /MTx and building as a DLL or linking against anything that uses the dynamic runtime will result in errors from crtdbg.h in Debug builds (and god knows what in Release builds) as g3logger and the executable or DLL using it will effectively use different heaps.

AndreasSchoenle replied:
Creating a static library may well have value. But linking it to the static runtime (which the /MT switches do) will make it unusable in most scenarios: It can then only be linked to an executable or DLL which uses the static runtime as well, which is rare. 
The reason is that it is dangerous at best in MSVC to mix DLL and static runtime or to use several modules (DLLs and Executables) all using the static runtime. Each independent module (DLL or EXE) linking the static runtime introduces its own heap and thus anything allocated dynamically within the module and destroyed outside will lead to heap corruption. 

Therefore all projects I know use the dynamic runtime which is the CMake default.

AndreasSchoenle replied:
Removed the commented out lines.

AndreasSchoenle commented on Options.cmake:
```diff
@@ -92,22 +100,13 @@ IF (MSVC OR MINGW)
-
-# Option for adding windows shared library
-    option (ADD_BUILD_WIN_SHARED  "Build shared library on Windows" OFF)
-       IF(ADD_BUILD_WIN_SHARED)
```
I have added the G3_SHARED_LIB option as was proposed in another pull request. I can rename it to ADD_BUILD_WIN_SHARED but as we are not adding but it but rather build as either shared or static I thought the name more appropriate

AndreasSchoenle commented on appveyor.yml:
```diff
@@ -18,7 +18,7 @@ install:
 
 build:
-  project: g3log\build\g3log.sln
+  project: c:\projects\g3log\build\g3log.sln
```
In fact the first option did not work for me. I put the absolute path to make it independent of any change of working directory in the script.

AndreasSchoenle commented:
I overlooked #167. Will not look into it for now and disable the test to make sure it does not show as an error any longer.

KjellKod commented on Options.cmake:
```diff
@@ -61,13 +63,21 @@ option (ENABLE_FATAL_SIGNALHANDLING
+# Option for building as a static or shared library in all platforms
+option (G3_SHARED_LIB  "Build shared library" ON)
+IF(G3_SHARED_LIB)
+   MESSAGE("Building shared library") 
```
MESSAGE("-DG3_SHARED_LIB=ON\tBuild shared library")

KjellKod commented on Options.cmake:
```diff
@@ -61,13 +63,21 @@ option (ENABLE_FATAL_SIGNALHANDLING
+IF(G3_SHARED_LIB)
+   MESSAGE("Building shared library") 
+ELSE() 
+   MESSAGE("Building static library") 
```
MESSAGE("-DG3_SHARED_LIB=OFF\tBuild static library")

KjellKod commented:
@AndreasSchoenle please see my pull request to your pull request. I think that addresses all of the comments that I had (maybe not all the questions). If it looks good please merge with your pull request and then I think this is good to go: https://github.com/AbberiorInstruments/g3log/pull/2

AndreasSchoenle commented on CMakeLists.txt:
```diff
@@ -49,10 +49,12 @@
 ENABLE_LANGUAGE(CXX)
 
-if(NOT CMAKE_BUILD_TYPE)
+if(NOT CMAKE_BUILD_TYPE AND NOT (MSVC_IDE OR XCODE))
```
Those are the multi-configuration generators I know - for those the setting is meaningless and the CMAKE_CONFIGURATION_TYPES variable holds those configurations being generated for the IDE

AndreasSchoenle commented on test_unit/test_io.cpp:
```diff
@@ -637,38 +637,38 @@ TEST(CustomLogLevels, AddANonFatal__DidtAddItToEnabledValue) {
 TEST(DynamicLogging, DynamicLogging_IS_ENABLED) {
```
I merged your pull request. Thanks!

AndreasSchoenle commented on Build.cmake:
```diff
@@ -93,20 +94,32 @@ ENDIF()
+      ENDIF()
+      ADD_LIBRARY(${G3LOG_LIBRARY} SHARED ${SRC_FILES})
+   ELSE()
+      ADD_LIBRARY(${G3LOG_LIBRARY} STATIC ${SRC_FILES})
```
The list APPEND is IMHO necessary, even on linux. If it did install, maybe it was also left over in the source directory and thus added through the file(GLOB command earlier? 
GLOB will not recursively add files, so even during an in-source build (in a build subdirectory as you do on linux) it should not get added.

AndreasSchoenle commented on Build.cmake:
```diff
@@ -120,13 +133,6 @@ ENDIF()
    ENDIF()
 
-   IF(NOT ${ADD_BUILD_WIN_SHARED}.x STREQUAL ".x")
-      IF(NOT(CMAKE_VERSION LESS 3.4) AND MSVC)
```
With my CMake it does work. I added test messages inside the if statement and it seemed to do what it was supposed to.

KjellKod commented:
@AndreasSchoenle awesome work and great explanations. 

👍

mobileben commented:
Excellent, thanks for this. This builds the static libraries issue that I had opened! I've successfully built the static libs with this.

KjellKod commented on src/g3log/loglevels.hpp:
```diff
@@ -90,14 +90,10 @@ namespace g3 {
-      INFO {g3::kInfoValue, {"INFO"}},
-      WARNING {g3::kWarningValue, {"WARNING"}},
-      FATAL {g3::kFatalValue, {"FATAL"}};
+const LEVELS G3LOG_DEBUG{g3::kDebugValue, {"DEBUG"}},
```
Ouch. We missed this. It’s broken!

KjellKod commented on src/g3log/loglevels.hpp:
```diff
@@ -90,14 +90,10 @@ namespace g3 {
-      WARNING {g3::kWarningValue, {"WARNING"}},
-      FATAL {g3::kFatalValue, {"FATAL"}};
+const LEVELS G3LOG_DEBUG{g3::kDebugValue, {"DEBUG"}},
+   INFO {g3::kInfoValue, {"INFO"}},
```
This is now handled on the cpp side : https://github.com/KjellKod/g3log/blob/master/src/loglevels.cpp#L22
